lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h

```diff
@@ -0,0 +1,81 @@
+#ifndef LIB_ANALYSIS_MULDEPTHANALYSIS_MULDEPTHANALYSIS_H_
+#define LIB_ANALYSIS_MULDEPTHANALYSIS_MULDEPTHANALYSIS_H_
+
+#include "mlir/include/mlir/Analysis/DataFlow/SparseAnalysis.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Diagnostics.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"    // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"        // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace mul_depth_analysis {
+
+class MulDepth {
+ public:
+  MulDepth() : value(std::nullopt) {}
+  MulDepth(int64_t value) : value(value) {}
+  ~MulDepth() = default;
+
+  bool isInitialized() const { return value.has_value(); }
+
+  const int64_t &getValue() const {
+    assert(isInitialized());
+    return *value;
+  }
+
+  bool operator==(const MulDepth &rhs) const { return value == rhs.value; }
+
+  static MulDepth join(const MulDepth &lhs, const MulDepth &rhs) {
+    if (!lhs.isInitialized()) return rhs;
+    if (!rhs.isInitialized()) return lhs;
+    return MulDepth{lhs.getValue() > rhs.getValue() ? lhs.getValue()
+                                                    : rhs.getValue()};
+  }
+
+  void print(raw_ostream &os) const { os << value; }
```

Comment:
Nit: when debugging there can be many printed values of various analysis, and it's hard to tell which is which without a name.

```suggestion
  void print(raw_ostream &os) const { os << "MulDepth(" << value << ")"; }
```